### ISSUE TITLE
re-activating Source Group makes the project wanted for fully reindex

### ISSUE THREAD

**Issue by bam80** [CoatiSoftware/Sourcetrail#1023]
Even if no other was changed at all and re-activated group was indexed previously.
Not sure if it's by design or not, but it hurts since it might be large codebase which needs to be re-indexed.

**Comment by mlangkabel** [comment]
What programming language are you indexing?

**Comment by bam80** [comment]
all Source Groups are C++ with Compilation Database json.

**Comment by bam80** [comment]
Do I need to provide additional info, or its already reproducible?

**Comment by mlangkabel** [comment]
Sorry for the delay. I didn't manage to reproduce the issue. I created two source groups, each contains a single `.cpp` file, both of which contain the same single `.h` file. 

* If I have both source groups active and change one of the `.cpp` files and use "Updated files" mode for re-indexing, it only wants to re-index the changed source file.
* If I have both source groups active and change one of the `.h` files and use "Updated files" mode for re-indexing, it wants to re-index both source file.
* If I indexed with just one source group active and then re-enable the second source group and use "Updated files" mode for re-indexing, it only wants to re-index the missing source file.

So it looks like everything works as it is supposed to be. Can you describe a minimal sample project that lets us reproduce the issue?

**Comment by bam80** [comment]
I didn't change any source files, just played with "active" group's flag.

> If I indexed with just one source group active and then re-enable the second source group and use "Updated files" mode for re-indexing, it only wants to re-index the missing source file.

If it was indexed before with both groups active I expect on re-activation it should just load previous indexed data and not re-index it, considering there were no changes to the files. Is your behavior different?

About the project, I have one but it's not minimal at all. Let me know if you still need it.

**Comment by mlangkabel** [comment]
The reasoning behind this behavior is that one usually doesn't have the same source file in two different source groups. The only reason to have one file in two source groups is if you want it to be indexed with different compiler flags. Different compiler flags could change the content that the compiler "sees" of the file. And in that case it is correct behavior that Sourcetrail re-indexes the file whenever you enable either source group.

**Comment by bam80** [comment]
But I don't have the same file in different groups. And still it tries to re-index all the project (even with "Updated files").

**Comment by bam80** [comment]
Here is the project with two groups. I generated compile_commands.json with CMake but maybe it's not important:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<config>
    <source_groups>
        <source_group_2fa64a88-9e90-43d1-8dc3-fda20b4429f7>
            <build_file_path>
                <compilation_db_path>build/kwin/compile_commands.json</compilation_db_path>
            </build_file_path>
            <exclude_filters>
                <exclude_filter>**_autogen/**</exclude_filter>
            </exclude_filters>
            <indexed_header_paths>
                <indexed_header_path>build</indexed_header_path>
                <indexed_header_path>src</indexed_header_path>
            </indexed_header_paths>
            <name>kwin</name>
            <pch_flags>
                <use_compiler_flags>1</use_compiler_flags>
            </pch_flags>
            <status>enabled</status>
            <type>C/C++ from Compilation Database</type>
        </source_group_2fa64a88-9e90-43d1-8dc3-fda20b4429f7>
        <source_group_4d3cf352-2be7-437c-b3e3-ee4ba8b134f2>
            <build_file_path>
                <compilation_db_path>build/plasma-desktop/compile_commands.json</compilation_db_path>
            </build_file_path>
            <exclude_filters>
                <exclude_filter>**_autogen/**</exclude_filter>
            </exclude_filters>
            <indexed_header_paths>
                <indexed_header_path>build</indexed_header_path>
                <indexed_header_path>src</indexed_header_path>
            </indexed_header_paths>
            <name>plasma-desktop</name>
            <pch_flags>
                <use_compiler_flags>1</use_compiler_flags>
            </pch_flags>
            <status>disabled</status>
            <type>C/C++ from Compilation Database</type>
        </source_group_4d3cf352-2be7-437c-b3e3-ee4ba8b134f2>
        <source_group_e5a10043-8a97-45db-bf92-4421ca906dec>
            <build_file_path>
                <compilation_db_path>build/attica/compile_commands.json</compilation_db_path>
            </build_file_path>
            <exclude_filters>
                <exclude_filter>**_autogen/**</exclude_filter>
            </exclude_filters>
            <indexed_header_paths>
                <indexed_header_path>build</indexed_header_path>
                <indexed_header_path>src</indexed_header_path>
            </indexed_header_paths>
            <name>attica</name>
            <pch_flags>
                <use_compiler_flags>1</use_compiler_flags>
            </pch_flags>
            <status>disabled</status>
            <type>C/C++ from Compilation Database</type>
        </source_group_e5a10043-8a97-45db-bf92-4421ca906dec>
    </source_groups>
    <version>8</version>
</config>

**Comment by bam80** [comment]
I'm quite interested to give it a go, have you had a chance to reproduce it with the project above?

**Comment by mlangkabel** [comment]
No, I didn't get a change to try that. I dont have the code for the projects your `.srctrlprj` file references on my machine. What happens if you try the steps in my comment from June 29th? Does that work?

**Comment by bam80** [comment]
Seems I narrowed down the problem. After triggering `active` group flags, indexing dialog still suggests 0 for "Updated files", that's OK.
But if we just re-start the program, it wants updating everything from scratch! :(
Can you repeat that behavior on your side?

**Comment by mlangkabel** [comment]
Ah, you mean that if you re-activate a source group in your project settings, it tells you that the settings have changed and that it needs to re-index the project and there it has the "Mode" set to "All files". So if you would click "Start" now, it would index everything (including the files that were already indexed). 
But what if you change the mode here to "Updated files"? Does it also index everything then? Or does that mode only index the "new" files?

**Comment by bam80** [comment]
It wants to re-index everything even on "Updated files" after the program is restarted.
I think it should be reproducible even on your minimal test project.

**Comment by mlangkabel** [comment]
Hm, If I re-activate one of my source groups, quit Sourcetrail without re-indexing, restart Sourcetrail and click "Reindex" when reloading the project, the "All files" option is selected and it shows me "Source files to index: 2". But if I switch to "Updated files", it shows "Source files to index: 1". So I'm afraid, I still cannot reproduce the behavior you experience.

**Comment by bam80** [comment]
Could you paste here your test project?

**Comment by mlangkabel** [comment]
Sure, here it is: [CxxSmallTest.zip](https://github.com/CoatiSoftware/Sourcetrail/files/5084907/CxxSmallTest.zip)
